common/crypto/hash.cpp

```diff
@@ -25,23 +30,153 @@
+// HMAC Functions
+// -----------------------------------------------------------------
+static void _ComputeHMAC_(
+    const EVP_MD *hashfunc(void),
+    const ByteArray& message,
+    const ByteArray& key,
+    ByteArray& hmac)
+{
+    const EVP_MD *md = hashfunc();
+    hmac.resize(EVP_MD_size(md));
+
+    std::unique_ptr<HMAC_CTX, void (*)(HMAC_CTX*)> hmac_ctx(HMAC_CTX_new(), HMAC_CTX_free);
+    pdo::error::ThrowIfNull(hmac_ctx.get(), "invalid hmac context");
+
+    int ret;
+    ret = HMAC_Init_ex(hmac_ctx.get(), key.data(), key.size(), md, NULL);
+    pdo::error::ThrowIf<pdo::error::RuntimeError>(ret == 0, "hmac init failed");
+
+    ret = HMAC_Update(hmac_ctx.get(), message.data(), message.size());
+    pdo::error::ThrowIf<pdo::error::RuntimeError>(ret == 0, "hmac update failed");
+
+    ret = HMAC_Final(hmac_ctx.get(), hmac.data(), NULL);
+    pdo::error::ThrowIf<pdo::error::RuntimeError>(ret == 0, "hmac final failed");
+}
+
+void pcrypto::SHA256HMAC(
+    const ByteArray& message,
+    const ByteArray& key,
+    ByteArray& hmac)
+{
+    _ComputeHMAC_(EVP_sha256, message, key, hmac);
+}
+
+void pcrypto::SHA384HMAC(
+    const ByteArray& message,
+    const ByteArray& key,
+    ByteArray& hmac)
+{
+    _ComputeHMAC_(EVP_sha384, message, key, hmac);
+}
+
+void pcrypto::SHA512HMAC(
+    const ByteArray& message,
+    const ByteArray& key,
+    ByteArray& hmac)
+{
+    _ComputeHMAC_(EVP_sha512, message, key, hmac);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+//
+static void _ComputePasswordBasedKeyDerivation_(
+    const EVP_MD *hashfunc(void),
+    const std::string& password,
+    const ByteArray& salt,
+    const unsigned int iterations,
+    ByteArray& key)
+{
+    const EVP_MD *md = hashfunc();
+    key.resize(EVP_MD_size(md));
+
+    int ret;
+    ret = PKCS5_PBKDF2_HMAC(
+        password.c_str(), password.size(),
+        salt.data(), salt.size(),
+        iterations, md,
+        key.size(), key.data());
+    pdo::error::ThrowIf<pdo::error::RuntimeError>(ret == 0, "password derivation failed");
+}
+
+void pcrypto::SHA512PasswordBasedKeyDerivation(
+    const std::string& password,
+    const ByteArray& salt,
+    ByteArray& key)
+{
+    _ComputePasswordBasedKeyDerivation_(EVP_sha512, password, salt, pcrypto::PBDK_Iterations, key);
+}
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// Compute SHA256 hash of message.data()
+// returns ByteArray containing raw binary data
+ByteArray pcrypto::ComputeMessageHash(const ByteArray& message)
+{
+    ByteArray hash(SHA256_DIGEST_LENGTH);
+    pcrypto::SHA256Hash(message, hash);
+    return hash;
+}  // pcrypto::ComputeMessageHash
+
+// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+// Compute SHA256-based HMAC of message.data()
+// returns ByteArray containing raw binary data
+ByteArray pcrypto::ComputeMessageHMAC(const ByteArray& key, const ByteArray& message)
```

Comment:
Given that we now expose the SHA xyz hash and hmac functions, I'm wondering what value we get from these generic ComputeMessageHash and HMAC.
My answer is: only backward compatibility, to avoid renaming these generic functions elsewhere.
However, going forward, it seems that we want to be specific about what we use.

Reply:
used in the python module.